Code/src/main/java/nl/utwente/group10/haskell/expr/Function.java

```diff
@@ -0,0 +1,167 @@
+            return HindleyMilner.makeVariable(); // TODO Make sure that this is correctly unified and/or make this editable.
+        }
+
+        @Override
+        public String toHaskell() {
+            return String.format("%s%d", Function.ARG_PREFIX, this.getPosition());
+        }
+
+        @Override
+        public String toString() {
+            return String.format("%s%d", Function.ARG_PREFIX, this.getPosition());
+        }
+    }
+
+    /** The prefix for function argument names. */
+    protected static final String ARG_PREFIX = "arg";
+
+    /** The arguments for this function. */
+    private List<FunctionArgument> arguments;
+
+    /** The expression that forms the base of this function. */
+    private Optional<Expr> expr;
+
+    /**
+     * @param expr The expression that forms the base of this function.
+     */
+    public Function(Expr expr) {
+        this.arguments = new ArrayList<>();
+        this.expr = Optional.ofNullable(expr);
+    }
+
+    /**
+     */
+    public Function() {
+        this(null);
+    }
+
+    /**
+     * Sets the expression for this function.
+     * @param expr The expression to set.
+     */
+    public void setExpr(Expr expr) {
+        this.expr = Optional.ofNullable(expr);
+    }
+
+    /**
+     * @return An ordered array of the arguments for this function.
+     */
+    public FunctionArgument[] getArguments() {
+        FunctionArgument[] args = new FunctionArgument[this.arguments.size()];
+        return this.arguments.toArray(args);
+    }
+
+    /**
+     * Adds a new argument to this function.
+     * @return The new argument.
+     */
+    public FunctionArgument addArgument() {
+        FunctionArgument arg = new FunctionArgument(this);
+        this.arguments.add(arg);
+        return arg;
+    }
+
+    /**
+     * Removes the given argument from this function. The argument is not removed from any expressions within this
+     * function.
+     * @param arg The argument to remove.
+     */
+    public void removeArgument(FunctionArgument arg) {
+        this.arguments.remove(arg);
+    }
+
+    @Override
+    public Type analyze(Env env, GenSet genSet) throws HaskellException {
+        Type type;
+
+        if (this.expr.isPresent()) {
+            type = this.expr.get().analyze(env, genSet);
+        } else {
+            type = HindleyMilner.makeVariable();
+
+            for (int i = this.arguments.size(); i > 0; i--) {
+                type = new FuncT(this.arguments.get(i-1).analyze(env, genSet), type);
+            }
+        }
+
+        return type;
+    }
+
+    @Override
+    public String toHaskell() {
+        StringBuilder out = new StringBuilder();
+
+        if (this.expr.isPresent()) {
+            out.append("\\");
+
+            for (FunctionArgument argument : this.arguments) {
+                out.append(" ").append(argument.toHaskell());
+            }
+
+            out.append(" = ");
+            out.append(this.expr.get().toHaskell());
+        }
```

Comment:
This is not entirely correct, lambda syntax in Haskell is `(\x -> x + 10)`, notice no space after backslash (although that may be allowed, I'm not sure) and arrow instead of equals sign.